cmd/otel-allocator/prehook/prehook.go

```diff
@@ -0,0 +1,56 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prehook
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/target"
+)
+
+const (
+	relabelConfigTargetFilterName = "relabel-config"
+)
+
+type Hook interface {
+	Apply(map[string]*target.TargetItem) map[string]*target.TargetItem
+	SetConfig(map[string][]*relabel.Config)
+	GetConfig() map[string][]*relabel.Config
+}
+
+type HookProvider func(log logr.Logger) Hook
+
+var (
+	registry = map[string]HookProvider{}
+)
+
+func New(name string, log logr.Logger) (Hook, error) {
+	if p, ok := registry[name]; ok {
+		return p(log.WithName("Prehook").WithName(name)), nil
```

Comment:
How does the no-op strategy work? I think what is going to happen for the default strategy of "no-op" it will never be registered, but set as the default which means that `New` will always return `nil, err` if someone doesn't specify the strategy. Could you add in something that registers the default "no-op" strategy?

Reply:
Thanks for the review! I think I have some cleanup to do for no-op, but in this current implementation I am following a previous suggestion to remove `no-op` entirely.  i.e. `allocation.New()` should return nil because then there will be no prehook and a nil check of allocator.filterFunction will indicate a no-op. So if a user specifices an unregistered strategt or doesn't specify a strategy , the action is the same - they will have a no-op w.r.t. filtering

Reply:
yep that's totally fine – as long as we clean up those no-op cases :)

Reply:
Does this function even need to return an error?  Can it just use the provided logger to log a warning and then return nil, which will be functionally equivalent to a `no-op`?